File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/mediator/ConcreteMediator.java
package jv.com.tencent.cain.design_pattern.mediator;

/**
 * @author cainjiang
 * @date 2020/8/13
 */
public class ConcreteMediator extends Mediator {
    private ConcreteColleague1 colleague1;
    private ConcreteColleague2 colleague2;

    @Override
    void send(Colleague colleague, String msg) {
        if (colleague instanceof ConcreteColleague1 && colleague2 != null) {
            colleague2.notify(msg);
        } else if (colleague instanceof ConcreteColleague2 && colleague != null) {
            colleague1.notify(msg);
        }
    }

    public ConcreteMediator setColleague1(ConcreteColleague1 colleague1) {
        this.colleague1 = colleague1;
        return this;
    }

    public ConcreteMediator setColleague2(ConcreteColleague2 colleague2) {
        this.colleague2 = colleague2;
        return this;
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/mediator/ConcreteMediator.kt
package kt.com.tencent.cain.desigin_pattern.mediator

/**
 * @author : jiangyu
 * @date   : 2020/8/13
 * @desc   : xxx
 */
class ConcreteMediator(var colleague1: Colleague? = null, var colleague2: Colleague? = null) : Mediator() {

    override fun send(sender: Colleague, msg: String) {
        when {
            sender is Colleague1 -> {
                colleague2?.notify(msg)
            }
            sender is Colleague2 -> {
                colleague1?.notify(msg)
            }
            else -> {
                println("nothing to do!")
            }
        }
    }
}